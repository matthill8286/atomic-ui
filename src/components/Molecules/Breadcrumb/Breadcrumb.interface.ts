import { MarginProps, PaddingProps, ThemeColors } from '@/types/theme'

export interface BreadcrumbPath {
  name: string
  link: string
  isRouterLink?: boolean
}

export type BreadcrumbProps = {
  className?: string
  loading?: boolean
  homeLink?: string
  paths: BreadcrumbPath[]
  hideLastElement?: boolean
  isHomeRouterLink?: boolean
  homeValue?: string
  textColor?: ThemeColors
  iconColor?: ThemeColors
  linkColor?: ThemeColors
  decorationColor?: ThemeColors
} & MarginProps &
  PaddingProps

export interface StyledLiProps {
  isLastButOne: boolean
}

export type StyledBreadcrumbProps = {
  className?: string
  isLoading: boolean
} & MarginProps &
  PaddingProps
