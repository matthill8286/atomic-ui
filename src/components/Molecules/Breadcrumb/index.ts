export * from './Breadcrumb'
export * from './Breadcrumb.interface'
