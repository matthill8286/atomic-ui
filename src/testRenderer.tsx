import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
