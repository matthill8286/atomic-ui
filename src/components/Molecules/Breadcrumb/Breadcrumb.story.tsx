import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { Breadcrumb } from './Breadcrumb'
import { BreadcrumbPath } from '@/components/Molecules/Breadcrumb/Breadcrumb.interface'

const paths: BreadcrumbPath[] = [
  {
    name: 'Dashboard',
    link: '/dashboard',
    isRouterLink: true,
  },
  {
    name: 'Learning Asset',
    link: '/learning-asset/3215',
    isRouterLink: true,
  },
  {
    name: 'Learning Asset: Watch about Genomes',
    link: '/learning-asset/3215-watch-about-genomes',
    isRouterLink: true,
  },
]

const knobs = () => ({
  margin: text('margin', 'lg sm'),
  padding: text('padding', 'lg sm'),
})

storiesOf('Design System/Molecules/Breadcrumb', module)
  .add('Default', () => (
    <Breadcrumb {...knobs()} paths={paths} hideLastElement={boolean('Hide Last Element', false)} />
  ))
  .add('Default as Skeleton', () => <Breadcrumb loading paths={[]} />)
