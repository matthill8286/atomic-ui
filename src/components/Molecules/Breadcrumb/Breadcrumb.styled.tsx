import { Icon } from '@/components/Atoms/Icon'
import { Link } from '@/components/Atoms/Link'
import { skeleton } from '@/styles/animation'
import { media } from '@/styles/media'
import { margin, padding } from '@/styles/sc-shared-functions'
import { css, styled } from '@/styles/styled'
import { StyledBreadcrumbProps, StyledLiProps } from './Breadcrumb.interface'

export const StyledBreadcrumb = styled.div<StyledBreadcrumbProps>`
  ${margin}
  ${padding}

  text-indent: ${({ isLoading }) => isLoading && '-9999px'};
  ${({ isLoading }) => isLoading && skeleton}
  min-height: ${({ theme }) => theme.font.lineHeight.sm};
`

export const StyledUl = styled.ul`
  display: flex;
  margin: 0;
  padding: 0;
`

export const StyledLi = styled.li<StyledLiProps>`
  display: inline-block;
  margin-right: ${({ theme }) => theme.spacing.base.xs};
  color: ${({ theme }) => theme.color.grey4};
  ${({ isLastButOne }) =>
    !isLastButOne
      ? css`
          display: none;
          ${media.xs} {
            display: inline-block;
          }
        `
      : css`
          ${media.md} {
            ${Icon} {
              & > svg {
                transform: rotate(0deg);
              }
            }
          }
        `}
`

export const StyledHomeLink = styled(Link)`
  ${Icon} {
    & > svg {
      position: relative;
      ${({ theme }) => css`
        top: ${theme.spacing.base.xxs};
        height: ${theme.spacing.base.sm};
        width: ${theme.spacing.base.sm};
      `}
    }
  }

  :hover {
    ${Icon} {
      & > svg {
        :hover {
          fill: ${({ theme }) => theme.color.grey4};
        }
      }
    }
  }
`
