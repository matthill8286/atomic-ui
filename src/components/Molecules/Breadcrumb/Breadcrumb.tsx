import * as React from 'react'
import { Icon } from '@/components/Atoms/Icon'
import { Link } from '@/components/Atoms/Link'
import { IconForwardSlash, StyleguideHomeVs } from '@excelwithbusiness/webmobile-svg-library'
import { BreadcrumbPath, BreadcrumbProps } from './Breadcrumb.interface'
import { StyledBreadcrumb, StyledHomeLink, StyledLi, StyledUl } from './Breadcrumb.styled'
import { useWindowDimensions } from '@/components/Helper'

const _renderSEOMeta = (paths: BreadcrumbPath[], homeLink: string): JSX.Element => {
  const breadcrumbSeoPaths: BreadcrumbPath[] = [...paths]

  const homeBreadcrumbPath: BreadcrumbPath = {
    name: 'home',
    link: homeLink,
  }

  breadcrumbSeoPaths.unshift(homeBreadcrumbPath)

  const meta = {
    '@context': 'https://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement: breadcrumbSeoPaths.map((seoPath, i) => ({
      '@type': 'ListItem',
      position: i + 1,
      name: seoPath.name,
      item: seoPath.link + (i === breadcrumbSeoPaths.length - 1 ? '#top' : ''),
    })),
  }

  return (
    <script type="application/ld+json" dangerouslySetInnerHTML={{ __html: JSON.stringify(meta) }} />
  )
}

export const Breadcrumb: React.FC<BreadcrumbProps> = ({
  className,
  margin,
  padding,
  paths,
  loading = false,
  homeLink = '/',
  hideLastElement = false,
  isHomeRouterLink = true,
  homeValue = 'Home',
  textColor = 'grey5',
  linkColor = 'grey5',
  iconColor = 'grey5',
  decorationColor,
  ...otherProps
}): JSX.Element => {
  const { breakpointName } = useWindowDimensions()
  const smallScreens = ['xs', undefined].includes(breakpointName)
  const visiblePaths = hideLastElement || smallScreens ? [...paths].slice(0, -1) : paths

  return (
    <>
      {_renderSEOMeta(paths, homeLink)}
      <StyledBreadcrumb
        className={className}
        margin={margin}
        padding={padding}
        isLoading={loading}
        data-test="filtered-breadcrumb-ul"
        data-paths-count={paths.length}
        {...otherProps}>
        <StyledUl>
          <StyledLi isLastButOne={false}>
            <StyledHomeLink
              inline
              {...(isHomeRouterLink ? { to: homeLink } : { href: homeLink })}
              fontSize="xs"
              underline={false}
              iconLeft={
                <Icon height="sm" width="sm" color={iconColor}>
                  {!loading && <StyleguideHomeVs fill={iconColor} />}
                </Icon>
              }
            />
          </StyledLi>
          {visiblePaths &&
            visiblePaths.map((item, index) => {
              const isLast = visiblePaths.length - 1 === index && !hideLastElement
              const isLastButOne = visiblePaths.length - 2 === index
              return (
                <StyledLi key={item.name} isLastButOne={isLastButOne}>
                  <Link
                    fontSize="sm"
                    inline
                    underline={false}
                    disabled={!smallScreens && isLast}
                    decorationColor={decorationColor}
                    iconLeft={<Icon color={iconColor}>{!loading && <IconForwardSlash />}</Icon>}
                    {...(item.isRouterLink ? { to: item.link } : { href: item.link })}>
                    {item.name}
                  </Link>
                </StyledLi>
              )
            })}
        </StyledUl>
      </StyledBreadcrumb>
    </>
  )
}

Breadcrumb.displayName = 'Breadcrumb'
