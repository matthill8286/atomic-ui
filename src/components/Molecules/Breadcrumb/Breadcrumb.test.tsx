import React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { Breadcrumb } from './Breadcrumb'

const paths = [
  {
    name: 'Computer & BÃ¼ro',
    link: '',
  },
  {
    name: 'Drucker & Scanner',
    link: '',
  },
  {
    name: 'Multifunktionsdrucker',
    link: '',
  },
]

describe('Heading', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(<Breadcrumb paths={paths} />)
    expect(tree).toMatchSnapshot()
  })
})
