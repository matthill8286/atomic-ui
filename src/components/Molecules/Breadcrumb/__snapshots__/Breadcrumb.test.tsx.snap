// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders correctly 1`] = `
Array [
  @media (min-width:752px) {

}

@media (min-width:752px) {

}

<script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"home\\",\\"item\\":\\"/\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"Computer & Büro\\",\\"item\\":\\"\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":3,\\"name\\":\\"Drucker & Scanner\\",\\"item\\":\\"\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":4,\\"name\\":\\"Multifunktionsdrucker\\",\\"item\\":\\"#top\\"}]}",
      }
    }
    type="application/ld+json"
  />,
  .c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 svg {
  width: 16px;
  height: 16px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  fill: #000000;
}

.c7 svg g,
.c7 svg p {
  fill: #000000;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 svg {
  width: 25px;
  height: 25px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  fill: #000000;
}

.c9 svg g,
.c9 svg p {
  fill: #000000;
}

.c10 {
  color: #666666;
  font-size: 16px;
  line-height: 24px;
  font-family: OswaldLight,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c3::after {
  content: '';
  position: absolute;
  display: none;
  bottom: -2px;
  height: 1px;
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c3:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #666666;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c8::after {
  content: '';
  position: absolute;
  display: none;
  bottom: -2px;
  height: 1px;
  background-color: #666666;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c8:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #E5E5E5;
  opacity: 0.4;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: initial;
  position: relative;
  padding: 16px;
}

.c13::after {
  content: '';
  position: absolute;
  display: none;
  bottom: -2px;
  height: 1px;
  background-color: #E5E5E5;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c5 {
  margin-left: 0px;
  margin-right: 4px;
}

.c5 svg {
  height: 16px;
  width: 16px;
}

.c11 {
  display: block;
  color: #666666;
}

.c0 {
  min-height: 24px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
}

.c2 {
  display: inline-block;
  display: none;
}

.c12 {
  display: inline-block;
}

.c12 .c6 > svg {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c4 .c6 > svg {
  position: relative;
  top: 4px;
  height: 16px;
  width: 16px;
}

.c4:hover .c6 > svg:hover {
  fill: #999999;
}

@media (min-width:752px) {
  .c2 {
    display: inline-block;
  }
}

@media (min-width:752px) {
  .c12 .c6 > svg {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
}

<div
    className="c0"
    data-paths-count={3}
    data-test="mms-breadcrumb-v2-ul"
  >
    <ul
      className="c1"
    >
      <li
        className="c2"
      >
        <a
          aria-label="home-link"
          className="c3 c4"
          color="grey4"
          href="/"
          target="_self"
        >
          <span
            className="c5"
            color="grey4"
          >
            <div
              className="c6 c7"
              color="black"
              height="sm"
              width="sm"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21.71 10.71l-7.42-7.42a1 1 0 00-.7-.29h-3.18a1 1 0 00-.7.29l-7.42 7.42a1 1 0 00-.29.7v1.09a.5.5 0 00.5.5H4v7a1 1 0 001 1h5a1 1 0 001-1v-4.75a.25.25 0 01.25-.25h1.5a.25.25 0 01.25.25V20a1 1 0 001 1h5a1 1 0 001-1v-7h1.5a.5.5 0 00.5-.5v-1.09a1 1 0 00-.29-.7zM18 19h-3v-5a1 1 0 00-1-1h-4a1 1 0 00-1 1v5H6V9.83L10.83 5h2.34L18 9.83z"
                />
              </svg>
            </div>
          </span>
        </a>
      </li>
      <li
        className="c2"
      >
        <a
          className="c8"
          color="grey5"
          href=""
          target="_self"
        >
          <span
            className="c5"
            color="grey5"
          >
            <div
              className="c6 c9"
              color="black"
            >
              <svg
                height={306}
                width={306}
              >
                <path
                  d="M94.35 0l-35.7 35.7L175.95 153 58.65 270.3l35.7 35.7 153-153z"
                />
              </svg>
            </div>
          </span>
          <span
            className="c10 c11"
            color="grey5"
            fontSize="sm"
          >
            Computer & Büro
          </span>
        </a>
      </li>
      <li
        className="c12"
      >
        <a
          className="c8"
          color="grey5"
          href=""
          target="_self"
        >
          <span
            className="c5"
            color="grey5"
          >
            <div
              className="c6 c9"
              color="black"
            >
              <svg
                height={306}
                width={306}
              >
                <path
                  d="M94.35 0l-35.7 35.7L175.95 153 58.65 270.3l35.7 35.7 153-153z"
                />
              </svg>
            </div>
          </span>
          <span
            className="c10 c11"
            color="grey5"
            fontSize="sm"
          >
            Drucker & Scanner
          </span>
        </a>
      </li>
      <li
        className="c2"
      >
        <a
          className="c13"
          color="grey5"
          target="_self"
        >
          <span
            className="c5"
            color="grey5"
          >
            <div
              className="c6 c9"
              color="black"
            >
              <svg
                height={306}
                width={306}
              >
                <path
                  d="M94.35 0l-35.7 35.7L175.95 153 58.65 270.3l35.7 35.7 153-153z"
                />
              </svg>
            </div>
          </span>
          <span
            className="c10 c11"
            color="grey5"
            fontSize="sm"
          >
            Multifunktionsdrucker
          </span>
        </a>
      </li>
    </ul>
  </div>,
]
`;
